### ISSUE TITLE
Reproducibility of results 

### ISSUE THREAD

**Issue by GigiProSerio** [DLR-RM/stable-baselines3#369]
I'll post it as a question as I am not quite sure that it is a bug.  I have been experimenting for a while with the library in a custom environment for a school project and I am really interested in the reproducibility of the result. I have read the disclaimer in the documentation that reads that reproducible results are not guaranteed across multiple platforms or different versions of Pytorch. However, I have been trying to reproduce the results on different platforms and so far I have managed to obtain the same results on different platforms with TD3, while the results for A2C and PPO instead keep changing. In both cases, I used the same code for seeding ( posted beneath) , while also seeding the algorithms in the parameters. Also, the algorithms are run on CPU . I wanted to ask if the fact that I can reproduce TD3 but not A2C and PPO is intended or I am doing something wrong. For A2C it may be due to the multiple workers however I don't understand why the same happens for PPO.  In case this behavior is to be expected could I try to solve it by running my algorithms on a docker for reproducibility?  

`import torch \
import numpy as np\
import random\
import os \
os.environ['PYTHONASHSEED'] = '0' \
seed = 0 \
torch.manual_seed(seed)\
torch.use_deterministic_algorithms(True)\
np.random.seed(seed)\
random.seed(seed)\ `

**Comment by Miffyli** [comment]
Good to hear at TD3 works out as intended! You have done good job setting all the seeds :)

I'd point at the "difference between platforms" as the source of the issue. I am not 100% sure if Docker images would fix this, and to be on the safe side I'd try running the code on same exact setup and seeding multiple times.  If that does not match, then it sounds like we have a bug, in which case minimal code to reproduce would be very much appreciated!

PPO also uses multiple environments, and generally on-policy algorithms (A2C, PPO) are more sensitive on what is being sampled from the environment (they also require more samples). TD3 storing things to a replay buffer stabilizes the learning (not so sensitive to the most recent samples).

**Comment by araffin** [comment]
>However, I have been trying to reproduce the results on different platforms and so far I have managed to obtain the same results on different platforms with TD3, while the results for A2C and PPO instead keep changing.

Before trying on different platforms, did you manage to have reproducible results with PPO/A2C on a single platform?

**Comment by GigiProSerio** [comment]
Thanks a lot for your answers. Yes, the results are fully reproducible on the two platforms I tried. On pc I get the same results and on Google colab I get the same result. However, even with the same seeding, hyperparameters and both running on cpu the results are different for PPO and A2C on the two platforms. I also set the number of environments for A2C and PPO to just 1, but the results keep being different. Another thing I noted is that the code I posted before for the seeding seems to be superfluous to get reproducible results on the same platform. All I need for that is passing the seed to the algorithm. On different platforms, however, even with the seeding posted and passing the seed to the algorithm I can't get reproducible results. Do you think there is anything else I should try ?

**Comment by Miffyli** [comment]
> On different platforms, however, even with the seeding posted and passing the seed to the algorithm I can't get reproducible results. Do you think there is anything else I should try ?

I'd try setting up Docker images or at least environments (e.g. Conda envs) with _exactly_ the same Python and library versions. The only next step after that might be setting up proper virtual machines with exact setups. After those options I think you have covered your options.

Alas, this is the state of the ML as it stands, where even a slight difference in underlying system libraries can make a difference :(

**Comment by GigiProSerio** [comment]
Thanks a lot for the help. I have already checked that the versions of stable baselines and pytorch are the same. I'll try with a conda environment perfectly replicating all the libraries and versions and next a docker!
